### ISSUE TITLE
Circular Convolution and Contour

### ISSUE THREAD

**Issue by promach** [zju3dv/snake#76]
I am reading both [实例分割新思路Deep Snake.pdf](https://github.com/zju3dv/snake/files/5041438/Deep.Snake.pdf) and [知乎: 实例分割新思路: Deep Snake (CVPR20'Oral Paper)](https://zhuanlan.zhihu.com/p/134111177) 

1. Could anyone explain in more details about the mathematical relationship between circular convolution and contour ?

2. Why if the number of neighbor node equals 2, then we could define convolution kernel ?

3. How does the [following code snippet](https://github.com/zju3dv/snake/blob/master/lib/networks/snake/evolve.py#L96) works in optimizing contour coordinates loss equation ?

```
            for i in range(self.iter):
                py_pred = py_pred / snake_config.ro
                c_py_pred = snake_gcn_utils.img_poly_to_can_poly(py_pred)
                evolve_gcn = self.__getattr__('evolve_gcn'+str(i))
                py_pred = self.evolve_poly(evolve_gcn, cnn_feature, py_pred, c_py_pred, init['py_ind'])
                py_preds.append(py_pred)
```

![contour](https://user-images.githubusercontent.com/3324659/89647159-a8dc1380-d8ef-11ea-8788-fe45ea79f618.png)

![Circular Convolution](https://user-images.githubusercontent.com/3324659/89647196-bdb8a700-d8ef-11ea-9a63-c01b7c85a4c8.png)

**Comment by pengsida** [comment]
1. You could learn more about the convolution equation.
2. The size of convolution kernel is 3.
3. The code passes the contour features into snake network and deforms the contour.

**Comment by promach** [comment]
2. How is kernel_size of 3 related to number of neighbor node ?
3. What are the purposes of the variables `py_pred` and `c_py_pred` ?

**Comment by pengsida** [comment]
1. Two neighboring nodes + one central node.
2. `py_pred` is the initial contour, `c_py_pred` is the normalized initial contour.

**Comment by promach** [comment]
What does this line of code : [py_pred = self.evolve_poly(evolve_gcn, cnn_feature, py_pred, c_py_pred, init['py_ind'])](https://github.com/zju3dv/snake/blob/0d8b70598da9e96a86f9815a42ca21da3f62075f/lib/networks/snake/evolve.py#L70) try to achieve ?

**Comment by pengsida** [comment]
The code passes the contour features into snake network and deforms the contour using the output offsets.

**Comment by promach** [comment]
Do bear with me since I am new to this "snake" network stuff.

I am checking [evolve_poly()](https://github.com/zju3dv/snake/blob/0d8b70598da9e96a86f9815a42ca21da3f62075f/lib/networks/snake/evolve.py#L70) function.

Would you be able to describe briefly what `i_poly = i_it_poly * snake_config.ro + snake(init_input, adj).permute(0, 2, 1)` does OR add some comment to the code itself ?

```
    def evolve_poly(self, snake, cnn_feature, i_it_poly, c_it_poly, ind):
        if len(i_it_poly) == 0:
            return torch.zeros_like(i_it_poly)
        h, w = cnn_feature.size(2), cnn_feature.size(3)
        init_feature = snake_gcn_utils.get_gcn_feature(cnn_feature, i_it_poly, ind, h, w)
        c_it_poly = c_it_poly * snake_config.ro
        init_input = torch.cat([init_feature, c_it_poly.permute(0, 2, 1)], dim=1)
        adj = snake_gcn_utils.get_adj_ind(snake_config.adj_num, init_input.size(2), init_input.device)
        i_poly = i_it_poly * snake_config.ro + snake(init_input, adj).permute(0, 2, 1)
        return i_poly
```

**Comment by pengsida** [comment]
It deforms the contour using the output offsets.
